File: solutions/linked-list-node-count/linked-list-node-count.scala
object Main extends App {

  abstract class LinkedList
  case class Node(value: Int, next: LinkedList) extends LinkedList
  case class Tail(value: Int) extends LinkedList

  def getLengthOfLinkedList(list: LinkedList): Int = {
    def traverseList(remainder: LinkedList, length: Int): Int = remainder match {
      case node: Node => traverseList(node.next, length + 1)
      case _: Tail => length + 1
    }
    traverseList(list, 0);
  }

  val linkedList = new Node(
    0, new Node(
      1, new Node(
        2, new Node(
          3, new Tail(4)
        )
      )
    )
  )

  println(getLengthOfLinkedList(linkedList))
}

File: solutions/linked-list-node-count/LinkedListNodeCount.java
public class LinkedListNodeCount {

  
  public static void main(String[] args) {

    Node n = new Node("one");
    LinkedList newList = new LinkedList(n);
    newList.addToList("two");
    newList.addToList("three");
    System.out.println(newList.getSize());


  }

    static class Node {
        Node next;
        Object item;
        public Node(Object o) {
            this.item = o;
            this.next = null;
        }
        
        public void setNext(Node n) {
            this.next = n;
        }
        
    }
    
    static class LinkedList {
        Node head;
        int size;
        public LinkedList(Node n) {
            
            this.head = n;
            this.size = 1;
        }
        
        public void addToList(Object o) {
            Node n = new Node(o);
            Node curr = this.head;
            if (head == null) {
                this.head = n;
                this.size += 1;
            }
            else {
                while (curr.next != null) {
                    curr = curr.next;
                }
                curr.next = n;
                this.size += 1;
            }
        }
        
        public int getSize() {
            return this.size;
        }
        
        
        
    }
    
}
